source/ShopPanel.cpp

```diff
@@ -388,6 +388,21 @@ bool ShopPanel::KeyDown(SDL_Keycode key, Uint16 mod, const Command &command, boo
 		return SetScrollToTop();
 	else if(key == SDLK_END)
 		return SetScrollToBottom();
+	else if(key == 'k' || (key == 'p' && (mod & KMOD_SHIFT)))
+	{
+		bool allParked = true;
```

Comment:
I was mimicing MZ's code in PlayerInfoPanel::Draw() — do you want me to change that as well?

Reply:
@RisingLeaf This is what the diff would look like replacing the `&=` for-loop with composition. Are you happy with adding this to the PR?

```diff
diff --git a/source/ShopPanel.cpp b/source/ShopPanel.cpp
index 855f137c8..97a2e2020 100644
--- a/source/ShopPanel.cpp
+++ b/source/ShopPanel.cpp
@@ -390,14 +390,13 @@ bool ShopPanel::KeyDown(SDL_Keycode key, Uint16 mod, const Command &command, boo
                return SetScrollToBottom();
        else if(key == 'k' || (key == 'p' && (mod & KMOD_SHIFT)))
        {
-               bool allParked = true;
                const Ship *flagship = player.Flagship();
-               for(Ship *ship : playerShips)
-               {
-                       if(!ship->IsDisabled() && ship != flagship)
-                               allParked &= ship->IsParked();
-               }
-               for(Ship *ship : playerShips)
+               const set<const Ship *> selectedEscorts;
+               copy_if(playerShips.begin(), playerShips.end(), back_inserter(selectedEscorts),
+                       [&](const Ship *ship){ return ship != flagship; });
+               bool allParked = all_of(selectedEscorts.begin(), selectedEscorts.end(),
+                       [&](const Ship *ship){ return ship->IsParked() && !ship->IsDisabled(); });
+               for(const Ship *ship : playerShips)
                {
                        if(!ship->IsDisabled() && ship != flagship)
                                player.ParkShip(ship, !allParked);
```

I feel I ought to filter out the flagship because the flagship should not return either true or false (it is always technically false, so if we returned that value for it, `allParked` would always be false for any selection including the flagship, and you'd be unable to Unpark such a selection until the flagship was removed from the set by the user). Really the 'allParked' variable means 'ignoringTheFlagshipAreAllOtherShipsParked`, so filtering it out is apropriate.
I could always return true for the flagship, pretending the flagship was always parked, then any selection that included all parked ships + the flagship could be Unparked. If this was in a hot loop I might do that, but it is in response to user input measured in the tens to hundreds of thousands of microseconds; an extra iteration of the selected ship list should be unnoticable.

Reply:
I am pretty sure you don't need that copy_if and can reduce it to a single all_of.

Reply:
Just have the lambda return true if the ship is the flagship. Also you don't need an all_of you just need to check if one escort is not parked, hence any_of

Reply:
> I am pretty sure you don't need that copy_if and can reduce it to a single all_of.

As I described at length, you don't need it, but by not filtering out the flagship, the code is lying. If you still think that's OK, I will do it.

any_of has early return, I could flip the logic and use that, you're right.

Reply:
> Just have the lambda return true if the ship is the flagship.

That won't work. Then a selection of the flagship and only parked escorts will try to park them again. I have updated it to remove the copy_if. I tried to be cunning with the variable name to avoid the code lying!

Reply:
resolved